File: app/src/main/java/com/dskja/betterstreamflix/providers/FilmPalastProvider.kt
package com.dskja.betterstreamflix.providers

import android.util.Log
import com.tanasi.retrofit_jsoup.converter.JsoupConverterFactory
import com.dskja.betterstreamflix.adapters.AppAdapter
import com.dskja.betterstreamflix.extractors.Extractor
import com.dskja.betterstreamflix.models.Category
import com.dskja.betterstreamflix.models.Episode
import com.dskja.betterstreamflix.models.Genre
import com.dskja.betterstreamflix.models.Movie
import com.dskja.betterstreamflix.models.People
import com.dskja.betterstreamflix.models.Season
import com.dskja.betterstreamflix.models.TvShow
import com.dskja.betterstreamflix.models.Video
import com.dskja.betterstreamflix.utils.NetworkClient
import com.dskja.betterstreamflix.utils.TmdbUtils
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.coroutineScope
import okhttp3.OkHttpClient
import okhttp3.ResponseBody
import org.jsoup.nodes.Document
import org.jsoup.select.Elements
import retrofit2.Response
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import retrofit2.http.GET
import retrofit2.http.Headers
import retrofit2.http.Path
import retrofit2.http.Url
import java.security.cert.CertPathValidatorException
import java.util.concurrent.TimeUnit
import javax.net.ssl.SSLHandshakeException
import javax.net.ssl.SSLPeerUnverifiedException

object FilmPalastProvider : Provider {

    private const val TAG = "FilmPalastProvider"
    private val BASE_URL = "https://filmpalast.to/"
    override val baseUrl = BASE_URL
    override val name = "Filmpalast"
    override val logo = "$BASE_URL/themes/downloadarchive/images/logo.png"
    override val language = "de"

    @Volatile
    private var service = FilmpalastService.build()
    @Volatile
    private var usingUnsafeSsl = false

    private fun isSslFailure(error: Throwable): Boolean {
        var current: Throwable? = error
        while (current != null) {
            when (current) {
                is SSLHandshakeException,
                is SSLPeerUnverifiedException,
                is CertPathValidatorException -> return true
            }
            val message = current.message.orEmpty()
            if (
                message.contains("certificate", ignoreCase = true) ||
                message.contains("CertPath", ignoreCase = true) ||
                message.contains("SSL", ignoreCase = true) ||
                message.contains("pretending to be", ignoreCase = true)
            ) {
                return true
            }
            current = current.cause
        }
        return false
    }

    private suspend fun <T> withSslFallback(block: suspend (FilmpalastService) -> T): T {
        return try {
            block(service)
        } catch (e: Exception) {
            if (usingUnsafeSsl || !isSslFailure(e)) throw e
            Log.w(TAG, "SSL failure talking to filmpalast.to; retrying with permissive TLS", e)
            service = FilmpalastService.buildUnsafe()
            usingUnsafeSsl = true
            block(service)
        }
    }

    override suspend fun getHome(): List<Category> {
        val document = withSslFallback { it.getHome() }
        val featured = coroutineScope {
            document.select("div.headerslider ul#sliderDla li").map { li ->
                async {
                    val title = li.select("span.title.rb").text()
                    val href = li.select("a.moviSliderPlay").attr("href")
                    val id = href.substringAfterLast("/")
                    val posterSrc = li.select("a img").attr("src")
                    val fullPosterUrl = if (posterSrc.startsWith("/")) {
                        "https://filmpalast.to$posterSrc"
                    } else {
                        posterSrc
                    }
                    val overview = li.select("div.moviedescription").text()

                    val releaseYearText = li.select("span.releasedate b").text()

                    val imdbRatingText =
                        li.select("span.views b").lastOrNull()?.text()?.split("/")?.get(1)?.trim()
                    val rating = imdbRatingText?.toDoubleOrNull() ?: 0.0

                    val tmdbMovie = TmdbUtils.getMovie(title, language = language)

                    Movie(
                        id = id,
                        title = title,
                        overview = overview,
                        released = releaseYearText,
                        rating = rating,
                        poster = fullPosterUrl,
                        banner = tmdbMovie?.banner
                    )
                }
            }.awaitAll()
        }


        val main_content = document.select("div#content article").map { article ->
            val href = article.selectFirst("h2 a")?.attr("href") ?: ""
            val title = article.selectFirst("h2 a")?.text() ?: ""
            val posterSrc = article.selectFirst("a img")?.attr("src") ?: ""

            val fullPosterUrl = if (posterSrc.startsWith("/")) {
                "https://filmpalast.to$posterSrc"
            } else {
                posterSrc
            }

            val info = article.select("*").toInfo()

            Movie(
                id = href.substringAfterLast("/"),
                title = title,
                released = info.released,
                quality = info.quality,
                rating = info.rating ?: 0.0,
                poster = fullPosterUrl
            )
        }
        val tvShowsDocument = withSslFallback { it.getTvShowsHome() }
        val tvShows = tvShowsDocument.select("div#content article").map { article ->
            val href = article.selectFirst("h2 a")?.attr("href") ?: ""
            val title = article.selectFirst("h2 a")?.text() ?: ""
            val posterSrc = article.selectFirst("a img")?.attr("src") ?: ""

            val fullPosterUrl = if (posterSrc.startsWith("/")) {
                "https://filmpalast.to$posterSrc"
            } else {
                posterSrc
            }

            val info = article.select("*").toInfo()                
                TvShow(
                id = href.substringAfterLast("/"),
                    title = title,
                    released = info.released,
                    quality = info.quality,
                    rating = info.rating ?: 0.0,
                    poster = fullPosterUrl
                )
        }

        return listOf(
            Category(name = "Featured", list = featured),
            Category(name = "Filme", list = main_content),
            Category(name = "Serien", list = tvShows)
        )
    }


    override suspend fun search(query: String, page: Int): List<AppAdapter.Item> {
        if (query.isEmpty()) {
            val document = withSslFallback { it.getHome() }
            val genres = document.select("aside#sidebar section#genre ul li a").map { element ->
                val text = element.text()
                Genre(text, text)
            }
            return genres
        }
        var document = withSslFallback { it.searchNoPage(query) }

        if (page > 1){
            val paging = document.selectFirst("div#paging a.pageing.button-small.rb")
            if (paging != null){
                document = withSslFallback { it.search(query, page) }
            } else {
                return emptyList()
            }
        }

        val results = document.select("div#content article").map { article ->
            val href = article.selectFirst("h2 a")?.attr("href") ?: ""
            val title = article.selectFirst("h2 a")?.text() ?: ""
            val posterSrc = article.selectFirst("a img")?.attr("src") ?: ""

            val fullPosterUrl = if (posterSrc.startsWith("/")) {
                "https://filmpalast.to$posterSrc"
            } else {
                posterSrc
            }

            val info = article.select("*").toInfo()

            // Determine if it's a movie or TV show based on season/episode info in title
            val isTvShow = title.matches(Regex(".*S\\d+E\\d+.*"))
            
            if (isTvShow) {
                TvShow(
                    id = href.substringAfterLast("/"),
                    title = title,
                    released = info.released,
                    quality = info.quality,
                    rating = info.rating ?: 0.0,
                    poster = fullPosterUrl
                )
            } else {
            Movie(
                id = href.substringAfterLast("/"),
                title = title,
                released = info.released,
                quality = info.quality,
                rating = info.rating ?: 0.0,
                poster = fullPosterUrl
            )
            }
        }.distinctBy { 
            when (it) {
                is Movie -> it.id
                is TvShow -> it.id
            }
        }

        return results

    }



    override suspend fun getMovie(id: String): Movie {
        val relativeId = BASE_URL + "stream/" + id;
        val document = withSslFallback { it.getMoviePage(relativeId) }
        val title = document.selectFirst("h2")?.text() ?: ""
        val poster = document.selectFirst("img.cover2")?.attr("src")?.let {
            if (it.startsWith("http")) it else "${BASE_URL.removeSuffix("/")}$it"
        }
        val description = document.selectFirst("span[itemprop=description]")?.text()
        val rating = document.selectFirst("div#star-rate")?.attr("data-rating")?.toDoubleOrNull()
        val genres =
            document.select("ul#detail-content-list > li:has(p:matchesOwn(Kategorien, Genre)) a")
                .map { Genre(id = it.text().trim(), name = it.text().trim()) }
        val directors = document.select("ul#detail-content-list > li:has(p:matchesOwn(Regie)) a")
            .map { People(id = it.text().trim(), name = it.text().trim()) }
        val actors =
            document.select("ul#detail-content-list > li:has(p:matchesOwn(Schauspieler)) a")
                .map { People(id = it.text().trim(), name = it.text().trim()) }

        val tmdbMovie = TmdbUtils.getMovie(title, language = language)

        return Movie(
            id = id,
            title = title,
            poster = poster,
            banner = tmdbMovie?.banner,
            genres = tmdbMovie?.genres ?: genres,
            directors = directors,
            cast = actors.map { person ->
                val tmdbPerson = tmdbMovie?.cast?.find { it.name.equals(person.name, ignoreCase = true) }
                person.copy(image = tmdbPerson?.image)
            },
            rating = tmdbMovie?.rating ?: rating,
            overview = tmdbMovie?.overview ?: description,
            released = tmdbMovie?.released?.let { "${it.get(java.util.Calendar.YEAR)}" } ?: document.selectFirst("ul#detail-content-list > li:has(p:matchesOwn(Release)) a")?.text()?.trim(),
            runtime = tmdbMovie?.runtime,
            trailer = tmdbMovie?.trailer,
            imdbId = tmdbMovie?.imdbId
        )
    }

    override suspend fun getServers(id: String, videoType: Video.Type): List<Video.Server> {
        val relativeId = BASE_URL + "stream/" + id;
        val document = withSslFallback { it.getMoviePage(relativeId) }
        val servers = mutableListOf<Video.Server>()

        val serverBlocks = document.select("ul.currentStreamLinks")
        val keywords = listOf("bigwarp", "vinovo")
        for (block in serverBlocks) {
            val name = block.selectFirst("li.hostBg p.hostName")?.text()?.trim() ?: "Unbekannt"
            var linkElement = block.selectFirst("a[href]")
            var url = linkElement?.attr("href")?.trim()
            if (linkElement == null){
                linkElement = block.selectFirst("a[data-player-url]")
                url = linkElement?.attr("data-player-url")?.trim();
            }


            if (!url.isNullOrEmpty()) {
                val displayName = if (keywords.none { name.lowercase().contains(it) }) {
                    name
                } else {
                    "$name (VLC Only)"
                }
                servers.add(
                    Video.Server(
                        id = name.split(" ")[0],
                        name = displayName,
                        src = url
                    )
                )
            }
        }

        return servers
    }

    override suspend fun getVideo(server: Video.Server): Video {
        val response = withSslFallback { it.getRedirectLink(server.src) }
            .let { response -> response.raw() as okhttp3.Response }


        val videoUrl = response.request.url
        val link = when (server.name) {
            "VOE" -> {
                val baseUrl = "https://voe.sx"
                val path = videoUrl.encodedPath.trimStart('/')
                "$baseUrl/e/$path?"
            }

            else -> videoUrl.toString()
        }
        return Extractor.extract(link, server)
    }

    override suspend fun getMovies(page: Int): List<Movie> {
        val document = withSslFallback { it.getMovies(page) }
        val movies = document.select("div#content article").map { article ->
            val href = article.selectFirst("h2 a")?.attr("href") ?: ""
            val title = article.selectFirst("h2 a")?.text() ?: ""
            val posterSrc = article.selectFirst("a img")?.attr("src") ?: ""

            val fullPosterUrl = if (posterSrc.startsWith("/")) {
                "https://filmpalast.to$posterSrc"
            } else {
                posterSrc
            }

            val info = article.select("*").toInfo()

            Movie(
                id = href.substringAfterLast("/"),
                title = title,
                released = info.released,
                quality = info.quality,
                rating = info.rating ?: 0.0,
                poster = fullPosterUrl
            )
        }
        return movies
    }

    override suspend fun getTvShows(page: Int): List<TvShow> {
        val document = withSslFallback { it.getTvShows(page) }
        val shows = document.select("div#content article").map { article ->
            val href = article.selectFirst("h2 a")?.attr("href") ?: ""
            val title = article.selectFirst("h2 a")?.text() ?: ""
            val posterSrc = article.selectFirst("a img")?.attr("src") ?: ""

            val fullPosterUrl = if (posterSrc.startsWith("/")) {
                "https://filmpalast.to$posterSrc"
            } else {
                posterSrc
            }

            val info = article.select("*").toInfo()
                
                TvShow(
                id = href.substringAfterLast("/"),
                    title = title,
                    released = info.released,
                    quality = info.quality,
                    rating = info.rating ?: 0.0,
                    poster = fullPosterUrl
                )
            }
        return shows
    }

    override suspend fun getTvShow(id: String): TvShow {
        val relativeId = BASE_URL + "stream/" + id
        val document = withSslFallback { it.getTvShow(relativeId) }
        val title = document.selectFirst("h2")?.text() ?: ""
        val poster = document.selectFirst("img.cover2")?.attr("src")?.let {
            if (it.startsWith("http")) it else "${BASE_URL.removeSuffix("/")}$it"
        }
        val description = document.selectFirst("span[itemprop=description]")?.text()
        val rating = document.selectFirst("div#star-rate")?.attr("data-rating")?.toDoubleOrNull()
        val genres = document.select("ul#detail-content-list > li:has(p:matchesOwn(Kategorien, Genre)) a")
            .map { Genre(id = it.text().trim(), name = it.text().trim()) }
        val directors = document.select("ul#detail-content-list > li:has(p:matchesOwn(Regie)) a")
            .map { People(id = it.text().trim(), name = it.text().trim()) }
        val actors = document.select("ul#detail-content-list > li:has(p:matchesOwn(Schauspieler)) a")
            .map { People(id = it.text().trim(), name = it.text().trim()) }

        // Parse seasons and episodes
        val seasons = mutableListOf<Season>()
        
        val cleanedTitle = title.replace(Regex("""\s+S\d+E\d+.*""", RegexOption.IGNORE_CASE), "")
            .replace(Regex("""\s+S\d+.*""", RegexOption.IGNORE_CASE), "")
            .trim()
            
        val tmdbTvShow = TmdbUtils.getTvShow(cleanedTitle, language = language)

        document.select("div#staffelWrapper div.staffelWrapperLoop").forEachIndexed { seasonIndex, seasonBlock ->
            val seasonNumber = seasonIndex + 1
            val episodes = mutableListOf<Episode>()
            
            seasonBlock.select("ul.staffelEpisodenList li a.getStaffelStream").forEachIndexed { episodeIndex, episodeLink ->
                val epTitle = episodeLink.ownText().trim()
                val epHref = episodeLink.attr("href")
                val fullEpHref = if (epHref.startsWith("//")) {
                    "https:$epHref"
                } else {
                    epHref
                }
                val epId = fullEpHref.substringAfterLast("/")
                
                episodes.add(
                    Episode(
                        id = epId,
                        number = episodeIndex + 1,
                        title = epTitle
                    )
                )
            }
            
            seasons.add(
                Season(
                    id = "${id}_$seasonNumber",
                    number = seasonNumber,
                    episodes = episodes,
                    poster = tmdbTvShow?.seasons?.find { it.number == seasonNumber }?.poster
                )
            )
        }

        return TvShow(
            id = id,
            title = title,
            poster = poster,
            banner = tmdbTvShow?.banner,
            genres = tmdbTvShow?.genres ?: genres,
            directors = directors,
            cast = actors.map { person ->
                val tmdbPerson = tmdbTvShow?.cast?.find { it.name.equals(person.name, ignoreCase = true) }
                person.copy(image = tmdbPerson?.image)
            },
            rating = tmdbTvShow?.rating ?: rating,
            overview = tmdbTvShow?.overview ?: description,
            released = tmdbTvShow?.released?.let { "${it.get(java.util.Calendar.YEAR)}" } ?: document.selectFirst("ul#detail-content-list > li:has(p:matchesOwn(Release)) a")?.text()?.trim(),
            runtime = tmdbTvShow?.runtime,
            trailer = tmdbTvShow?.trailer,
            imdbId = tmdbTvShow?.imdbId,
            seasons = seasons
        )
    }

    override suspend fun getGenre(id: String, page: Int): Genre {
        val document = withSslFallback { it.getGenre(id, page) }

        val shows = document.select("div#content article").map { article ->
            val aTag = article.selectFirst("h2 a")
            val href = aTag?.attr("href").orEmpty()
            val title = aTag?.text().orEmpty()

            val posterSrc = article.selectFirst("a img")?.attr("src").orEmpty()
            val fullPosterUrl = if (posterSrc.startsWith("/")) {
                "https://filmpalast.to$posterSrc"
            } else {
                posterSrc
            }

            val info = article.select("*").toInfo()

            Movie(
                id = href.substringAfterLast("/"),
                title = title,
                released = info.released,
                quality = info.quality,
                rating = info.rating ?: 0.0,
                poster = fullPosterUrl
            )
        }

        return Genre(
            id = id,
            name = id.replaceFirstChar { it.uppercase() },
            shows = shows
        )
    }


    override suspend fun getEpisodesBySeason(seasonId: String): List<Episode> {
        val parts = seasonId.split("_")
        if (parts.size != 2) return emptyList()
        
        val showId = parts[0]
        val seasonNumber = parts[1].toIntOrNull() ?: return emptyList()
        
        val relativeId = BASE_URL + "stream/" + showId
        val document = withSslFallback { it.getTvShow(relativeId) }
        val title = document.selectFirst("h2")?.text() ?: ""

        val cleanedTitle = title.replace(Regex("""\s+S\d+E\d+.*""", RegexOption.IGNORE_CASE), "")
            .replace(Regex("""\s+S\d+.*""", RegexOption.IGNORE_CASE), "")
            .trim()

        val tmdbTvShow = TmdbUtils.getTvShow(cleanedTitle, language = language)
        val tmdbEpisodes = tmdbTvShow?.let { TmdbUtils.getEpisodesBySeason(it.id, seasonNumber, language = language) } ?: emptyList()
        
        val episodes = mutableListOf<Episode>()
        
        document.select("div#staffelWrapper div.staffelWrapperLoop").forEachIndexed { index, seasonBlock ->
            if (index + 1 == seasonNumber) {
                seasonBlock.select("ul.staffelEpisodenList li a.getStaffelStream").forEachIndexed { episodeIndex, episodeLink ->
                    val epTitle = episodeLink.ownText().trim()
                    val epHref = episodeLink.attr("href")
                    val fullEpHref = if (epHref.startsWith("//")) {
                        "https:$epHref"
                    } else {
                        epHref
                    }
                    val epId = fullEpHref.substringAfterLast("/")
                    val epNumber = episodeIndex + 1

                    val tmdbEp = tmdbEpisodes.find { it.number == epNumber }
                    
                    episodes.add(
                        Episode(
                            id = epId,
                            number = epNumber,
                            title = tmdbEp?.title ?: epTitle,
                            poster = tmdbEp?.poster,
                            overview = tmdbEp?.overview
                        )
                    )
                }
            }
        }
        
        return episodes
    }
    override suspend fun getPeople(id: String, page: Int): People {
        val url = "$BASE_URL/search/title/$id"
        val document = withSslFallback { it.getPeoplePage(url) }
        val name = document.selectFirst("h1")?.text() ?: ""
        val image = document.selectFirst("img.cover2")?.attr("src")?.let {
            if (it.startsWith("http")) it else "${BASE_URL.removeSuffix("/")}$it"
        }
        
        // Parse filmography (same structure as movies/series)
        val filmography = document.select("div#content article").map { article ->
            val href = article.selectFirst("h2 a")?.attr("href") ?: ""
            val title = article.selectFirst("h2 a")?.text() ?: ""
            val posterSrc = article.selectFirst("a img")?.attr("src") ?: ""

            val fullPosterUrl = if (posterSrc.startsWith("/")) {
                "https://filmpalast.to$posterSrc"
            } else {
                posterSrc
            }

            val info = article.select("*").toInfo()

            // Determine if it's a movie or TV show based on season/episode info in title
            val isTvShow = title.matches(Regex(".*S\\d+E\\d+.*"))
            
            if (isTvShow) {
                TvShow(
                    id = href.substringAfterLast("/"),
                    title = title,
                    released = info.released,
                    quality = info.quality,
                    rating = info.rating ?: 0.0,
                    poster = fullPosterUrl
                )
            } else {
                Movie(
                    id = href.substringAfterLast("/"),
                    title = title,
                    released = info.released,
                    quality = info.quality,
                    rating = info.rating ?: 0.0,
                    poster = fullPosterUrl
                )
            }
        }
        
        return People(
            id = id,
            name = name,
            image = image,
            filmography = filmography
        )
    }


    private fun Elements.toInfo() =
        this.mapNotNull { it.text().trim().takeIf { it.isNotEmpty() } }.let { textList ->
            val starCount = this.select("img[src*=star_on]").size
            val rating = starCount / 10.0

            object {
                val rating =
                    rating.takeIf { starCount > 0 }

                val quality = textList.find { it in listOf("HD", "SD", "CAM", "TS", "HDRip") }

                val released = textList.find { it.matches(Regex("\\d{4}")) }

            }
        }


    interface FilmpalastService {

        companion object {
            private fun client(trustAll: Boolean): OkHttpClient {
                val base = if (trustAll) NetworkClient.trustAll else NetworkClient.default
                return base.newBuilder()
                    .readTimeout(30, TimeUnit.SECONDS)
                    .connectTimeout(30, TimeUnit.SECONDS)
                    .build()
            }

            fun build(): FilmpalastService = create(client(trustAll = false))

            fun buildUnsafe(): FilmpalastService = create(client(trustAll = true))

            private fun create(client: OkHttpClient): FilmpalastService {
                val retrofit = Retrofit.Builder().baseUrl(BASE_URL)
                    .addConverterFactory(JsoupConverterFactory.create())
                    .addConverterFactory(GsonConverterFactory.create())
                    .client(client)
                    .build()
                return retrofit.create(FilmpalastService::class.java)
            }
        }

        @GET("movies/new/page/1")
        suspend fun getHome(): Document

        @GET("serien/view/page/1")
        suspend fun getTvShowsHome(): Document

        @GET
        suspend fun getMoviePage(@Url url: String): Document

        @GET
        suspend fun getTvShow(@Url url: String): Document

        @GET("movies/new/page/{page}")
        suspend fun getMovies(@Path("page") page: Int): Document

        @GET("serien/view/page/{page}")
        suspend fun getTvShows(@Path("page") page: Int): Document

        @GET("search/title/{query}/{page}")
        suspend fun search(@Path("query") query: String,@Path("page") page: Int): Document
        @GET("search/title/{query}")
        suspend fun searchNoPage(@Path("query") query: String): Document

        @GET
        @Headers("User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64)")
        suspend fun getRedirectLink(@Url url: String): Response<ResponseBody>

        @GET("search/genre/{genre}/{page}")
        suspend fun getGenre(@Path("genre") genre: String, @Path("page") page: Int): Document

        @GET
        suspend fun getPeoplePage(@Url url: String): Document

    }

}

File: ios/BetterStreamflix/Core/Providers/FilmPalastProvider.swift
import Foundation
import SwiftSoup

struct FilmPalastProvider: CatalogProvider {
    let id = "filmpalast"
    let name = "Filmpalast"
    let language = "de"
    let baseURL = URL(string: "https://filmpalast.to/")!

    func home() async throws -> [CategoryRow] {
        var rows: [CategoryRow] = []

        if let html = try? await HTTPClient.getHTML(
            url: URL(string: "movies/new/page/1", relativeTo: baseURL)!.absoluteURL,
            desktopUA: true,
            allowLenientTLS: true
        ),
           let doc = try? SwiftSoup.parse(html, baseURL.absoluteString) {
            let featured = try doc.select("div.headerslider ul#sliderDla li").array().compactMap { li -> MediaItem? in
                let title = try li.selectFirst("span.title.rb")?.text() ?? ""
                let href = try li.selectFirst("a.moviSliderPlay")?.attr("href") ?? ""
                let id = href.split(separator: "/").last.map(String.init) ?? ""
                guard !id.isEmpty, !title.isEmpty else { return nil }
                let poster = HTTPClient.absoluteURL(try li.selectFirst("a img")?.attr("src"), base: baseURL)
                return MediaItem(
                    id: id,
                    title: title,
                    posterURL: poster,
                    bannerURL: poster,
                    overview: try li.selectFirst("div.moviedescription")?.text(),
                    year: try li.selectFirst("span.releasedate b")?.text(),
                    kind: .movie,
                    providerHint: self.id
                )
            }
            if !featured.isEmpty {
                rows.append(CategoryRow(id: "featured", title: "Featured", items: featured, isFeatured: true))
            }

            let movies = try parseArticles(doc, kind: .movie)
            if !movies.isEmpty {
                rows.append(CategoryRow(id: "movies", title: "Filme", items: movies))
            }
        }

        if let tvHTML = try? await HTTPClient.getHTML(
            url: URL(string: "serien/view/page/1", relativeTo: baseURL)!.absoluteURL,
            referer: baseURL,
            desktopUA: true,
            allowLenientTLS: true
        ),
           let tvDoc = try? SwiftSoup.parse(tvHTML, baseURL.absoluteString) {
            let shows = try parseArticles(tvDoc, kind: .tvShow)
            if !shows.isEmpty {
                rows.append(CategoryRow(id: "series", title: "Serien", items: shows))
            }
        }

        if rows.isEmpty {
            throw ProviderError.parseFailed("Filmpalast unreachable")
        }
        return rows
    }

    func search(query: String) async throws -> [MediaItem] {
        let trimmed = query.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return [] }
        let encoded = trimmed.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed) ?? trimmed
        let url = URL(string: "search/title/\(encoded)", relativeTo: baseURL)!.absoluteURL
        let html = try await HTTPClient.getHTML(url: url, referer: baseURL, desktopUA: true, allowLenientTLS: true)
        let doc = try SwiftSoup.parse(html, baseURL.absoluteString)
        // Prefer path-based kind detection
        var items: [MediaItem] = []
        var seen = Set<String>()
        for article in try doc.select("div#content article").array() {
            let href = try article.selectFirst("h2 a")?.attr("href") ?? ""
            let id = href.split(separator: "/").last.map(String.init) ?? ""
            guard !id.isEmpty, seen.insert(id).inserted else { continue }
            let title = try article.selectFirst("h2 a")?.text() ?? id
            let poster = HTTPClient.absoluteURL(try article.selectFirst("a img")?.attr("src"), base: baseURL)
            let kind: MediaItem.Kind = href.contains("/serien/") || href.contains("/series/") ? .tvShow : .movie
            items.append(
                MediaItem(id: id, title: title, posterURL: poster, kind: kind, providerHint: self.id)
            )
        }
        if items.isEmpty {
            items = (try parseArticles(doc, kind: .movie)) + (try parseArticles(doc, kind: .tvShow))
        }
        return items
    }

    func detail(id: String, kind: MediaItem.Kind) async throws -> ShowDetail {
        let pathCandidates = [
            "stream/\(id).html",
            "stream/\(id)",
            id.hasSuffix(".html") ? "stream/\(id)" : "stream/\(id).html",
        ]
        var lastError: Error = ProviderError.emptyResponse
        for path in pathCandidates {
            guard let url = URL(string: path, relativeTo: baseURL)?.absoluteURL else { continue }
            do {
                let html = try await HTTPClient.getHTML(url: url, referer: baseURL, desktopUA: true, allowLenientTLS: true)
                let doc = try SwiftSoup.parse(html, baseURL.absoluteString)
                let title = try doc.selectFirst("h1")?.text()
                    ?? id.replacingOccurrences(of: "-", with: " ").capitalized
                let overview = try doc.selectFirst("div.detail-desc, #detail-desc, .seriesDescr")?.text()
                let poster = HTTPClient.absoluteURL(
                    try doc.selectFirst(".detail-cover img, .cover img, img[itemprop=image]")?.attr("src"),
                    base: baseURL
                )
                return ShowDetail(
                    id: id,
                    title: title,
                    overview: overview,
                    posterURL: poster,
                    bannerURL: poster,
                    year: nil,
                    rating: nil,
                    seasons: kind == .tvShow
                        ? [SeasonInfo(id: id, number: 1, title: "Episodes")]
                        : [],
                    kind: kind
                )
            } catch {
                lastError = error
            }
        }
        throw lastError
    }

    func episodes(showId: String, seasonId: String) async throws -> [EpisodeInfo] {
        let url = URL(string: "stream/\(showId).html", relativeTo: baseURL)!.absoluteURL
        let html = try await HTTPClient.getHTML(url: url, referer: baseURL, desktopUA: true, allowLenientTLS: true)
        let doc = try SwiftSoup.parse(html, baseURL.absoluteString)
        var episodes: [EpisodeInfo] = []
        for (idx, link) in try doc.select("a[href*=staffel], a[href*=episode], .episodelist a").array().enumerated() {
            let href = try link.attr("href")
            let title = try link.text().trimmingCharacters(in: .whitespacesAndNewlines)
            let epId = href.split(separator: "/").last.map(String.init) ?? "\(showId)-e\(idx + 1)"
            episodes.append(
                EpisodeInfo(
                    id: epId,
                    number: idx + 1,
                    title: title.isEmpty ? "Episode \(idx + 1)" : title
                )
            )
        }
        if episodes.isEmpty {
            episodes = [EpisodeInfo(id: showId, number: 1, title: "Play")]
        }
        return episodes
    }

    func streams(showId: String, seasonId: String?, episodeId: String?, detail: ShowDetail?) async throws -> [StreamSource] {
        let target = episodeId ?? showId
        let path = target.contains(".html") ? "stream/\(target)" : "stream/\(target).html"
        let url = URL(string: path, relativeTo: baseURL)!.absoluteURL
        let html = try await HTTPClient.getHTML(url: url, referer: baseURL, desktopUA: true, allowLenientTLS: true)
        let doc = try SwiftSoup.parse(html, baseURL.absoluteString)
        var sources: [StreamSource] = []

        for (idx, el) in try doc.select("ul.currentStreamLinks a, a.iconPlay, a[data-player-url], a[href*=voe], a[href*=streamtape], a[href*=vidoza]").array().enumerated() {
            let href = try el.attr("data-player-url").ifBlank(try el.attr("href"))
            guard var streamURL = HTTPClient.absoluteURL(href, base: baseURL) else { continue }
            let name = try el.text().trimmingCharacters(in: .whitespacesAndNewlines)
            if streamURL.host()?.contains("filmpalast") == true,
               let path = URLComponents(url: streamURL, resolvingAgainstBaseURL: false)?.path,
               path.contains("/e/") || path.contains("/v/") {
                // leave as-is; followRedirect will resolve
            }
            // VOE often needs voe.sx host
            if let host = streamURL.host(), host.contains("voe"), !host.contains("voe.sx") {
                streamURL = URL(string: "https://voe.sx\(streamURL.path)") ?? streamURL
            }
            sources.append(
                StreamSource(
                    id: "fp-\(idx)",
                    name: name.isEmpty ? "Host \(idx + 1)" : name,
                    url: streamURL,
                    headers: ["User-Agent": HTTPClient.desktopUserAgent, "Referer": baseURL.absoluteString],
                    resolveKind: .followRedirect
                )
            )
        }
        return sources.uniqued(by: \.url)
    }

    private func parseArticles(_ doc: Document, kind: MediaItem.Kind) throws -> [MediaItem] {
        var items: [MediaItem] = []
        var seen = Set<String>()
        for article in try doc.select("div#content article").array() {
            let href = try article.selectFirst("h2 a")?.attr("href") ?? ""
            let id = href.split(separator: "/").last.map(String.init) ?? ""
            guard !id.isEmpty, seen.insert(id).inserted else { continue }
            let title = try article.selectFirst("h2 a")?.text() ?? id
            let poster = HTTPClient.absoluteURL(try article.selectFirst("a img")?.attr("src"), base: baseURL)
            items.append(
                MediaItem(id: id, title: title, posterURL: poster, kind: kind, providerHint: self.id)
            )
        }
        return items
    }
}

private extension Array {
    func uniqued<T: Hashable>(by keyPath: KeyPath<Element, T>) -> [Element] {
        var seen = Set<T>()
        return filter { seen.insert($0[keyPath: keyPath]).inserted }
    }
}

private extension String {
    func ifBlank(_ fallback: String?) -> String {
        let trimmed = trimmingCharacters(in: .whitespacesAndNewlines)
        if trimmed.isEmpty { return fallback?.trimmingCharacters(in: .whitespacesAndNewlines) ?? "" }
        return trimmed
    }
}

private extension Optional where Wrapped == String {
    func ifBlank(_ fallback: String?) -> String? {
        guard let self else { return fallback }
        return self.ifBlank(fallback)
    }
}
